core/geometry/src/curve/internalContexts/geometry2d/ConstraintConstruction.ts

```diff
@@ -0,0 +1,271 @@
+        n++;
+    }
+    return n;
+  }
+  /** Replace zero radius circle constraints to point constraints. */
+  public convertZeroRadiusCirclesToThroughPoint(): void {
+    for (let i = 0; i < this._constraints.length; i++) {
+      const c = this._constraints[i];
+      if (c.constraintType === ConstraintType.tangentTo
+        && c.curve !== undefined
+        && c.curve instanceof UnboundedCircle2dByCenterAndRadius
+        && Geometry.isSmallMetricDistance(c.curve.radius)) {
+        const c1 = ConstraintConstruction.createThroughPoint(c.curve.center);
+        this._constraints[i] = c1;
+      }
+    }
+  }
+  /** Replace point passthrough to zero radius circle tangency. */
+  public convertThroughPointToZeroRadiusCircles(): void {
+    for (let i = 0; i < this._constraints.length; i++) {
+      const c = this._constraints[i];
+      if (c.constraintType === ConstraintType.throughPoint && c.point !== undefined) {
+        const c1 = ConstraintConstruction.createTangentTo(
+          UnboundedCircle2dByCenterAndRadius.createPointRadius(c.point, 0.0))
+        this._constraints[i] = c1;
+      }
+    }
+  }
+  /** Normalize all line perpendiculars. */
+  public normalizeLines(): void {
+    for (let i = 0; i < this._constraints.length; i++) {
+      const c = this._constraints[i];
+      if (c.constraintType === ConstraintType.tangentTo
+        && c.curve instanceof UnboundedLine2dByPointAndNormal) {
+        const newLine = c.curve.cloneNormalizedFromOrigin();
+        if (newLine !== undefined) {
+          const c1 = ConstraintConstruction.createTangentTo(newLine);
+          this._constraints[i] = c1;
+        }
+      }
+    }
+  }
+  /**
+   * Return
+   *   * 1 if the constraint curve is UnboundedLine2dByPointAndNormal
+   *   * 2 if the constraint curve is UnboundedCircle2dByCenterAndRadius
+   *   * 0 otherwise
+   * @param c the constraint.
+   */
+  public static tangentGeometryLineCircleType(c: ConstraintConstruction): number {
+    if (c.curve instanceof UnboundedCircle2dByCenterAndRadius)
+      return 2;
+    if (c.curve instanceof UnboundedLine2dByPointAndNormal)
+      return 1;
+    return 0;
+  }
+  /**
+   * Sort the given constraints with
+   * * Primary sort is on constraint type
+   * * For equal constraint type, apply ordering from tangentGeometryType function.
+   */
+  public static sortByConstraintTypeAndGeometryType(constraints: ConstraintConstruction[]) {
+    constraints.sort(
+      (a: ConstraintConstruction, b: ConstraintConstruction) => {
+        if (a.constraintType === b.constraintType) {
+          const ga = this.tangentGeometryLineCircleType(a);
+          const gb = this.tangentGeometryLineCircleType(b);
+          if (ga === gb)
+            return 0;
+          return ga < gb ? -1 : 1;
+        }
+        return (a.constraintType < b.constraintType) ? -1 : 1;
+      }
+    )
+  }
+  public constructConstrainedCircles(): ImplicitGeometryMarkup<UnboundedCircle2dByCenterAndRadius>[] | undefined {
```

Comment:
Please make sure to mention in the function doc that constraint set has to have 3 constraints.

Reply:
(Pushed 8/26)